### PR TITLE
[next] Parent Threading

### PR DESCRIPTION
**Pull request by wyattjoh**
## What does this PR do?

Introduces a new concept for storing thread information for a given parent/child relationship to drastically improve deep parent retrieval.

## 🚨 Caveat 🚨

This does introduce a technical limitation however, as [MongoDB limits the size of documents to 16 MB](https://docs.mongodb.com/manual/reference/limits/#bson-documents), it roughly equates to a given comment having ~20K ID's referencing a given comment in the upwards or downwards direction.

If for example, a comment is 100 levels deep, it can have a maximum of (20K - 100) direct replies to it. Safeguards have been added to Talk to ensure that this database limitation is not reached, but based on real world comment installations, you are pretty much guaranteed to not hit this limit.

### CODE REVIEW

File: src/core/server/graph/tenant/schema/schema.graphql [hunk 1]
```diff
@@ -624,11 +640,22 @@ type Comment {
   """
   status: COMMENT_STATUS!
 
+  """
+  parentCount is the number of direct parents for this Comment. Currently this
+  value is the same as depth.
```
Comment (alinhle) [inline src/core/server/graph/tenant/schema/schema.graphql:645]: When will this ever be different?

File: src/core/server/models/comment.ts [hunk 2]
```diff
@@ -273,6 +303,96 @@ export async function retrieveCommentRepliesConnection(
...
+  if (!comment.parent_id) {
+    return createConnection({
+      pageInfo: {
+        hasNextPage: false,
+        hasPreviousPage: false,
+      },
+    });
+  }
+
+  // TODO: (wyattjoh) maybe throw an error when the limit is zero?
```
Comment (alinhle) [inline src/core/server/models/comment.ts:331]: I'd throw if all it does is returning an empty connection.

File: src/core/server/models/comment.ts [hunk 3]
```diff
@@ -273,6 +303,96 @@ export async function retrieveCommentRepliesConnection(
...
+  // Fetch the subset of the comment id's that we are going to query for.
+  const parentIDSubset = parentIDs.slice(skip + 1, skip + 1 + limit);
+
+  // Retrieve the parents via the subset list.
+  const parents = await retrieveManyComments(mongo, tenantID, parentIDSubset);
+
+  // Loop over the list to ensure that none of the entries is null (indicating
+  // that there was a misplaced parent). We can assert the type here because we
+  // will throw an error and abort if one of the comments are null.
+  const nodes = parents.filter(parentComment => {
```
Comment (alinhle) [inline src/core/server/models/comment.ts]: `forEach` would also do :-)

**Review by alinhle — CHANGES_REQUESTED** [review]

File: src/core/server/models/comment.ts (hunk 2 above)
Comment (wyattjoh) [reply src/core/server/models/comment.ts:331]: An empty connection is still technically valid, albeit not super useful.

File: src/core/server/graph/tenant/schema/schema.graphql (hunk 1 above)
Comment (wyattjoh) [reply src/core/server/graph/tenant/schema/schema.graphql:645]: If we decide to change the count based on deleted/rejected/tombstoned comments

**Review by alinhle — APPROVED** [review]

**Pull request merged**